File: bsp/src/org/jetbrains/bsp/protocol/session/CancellableFuture.scala
package org.jetbrains.bsp
package protocol
package session

import java.util.concurrent.CompletableFuture

/**
 * JDK's CompletableFuture does not handle cancellation well.
 * When a future overrides a `cancel` method, it is lost
 * when doing transformations like `thenApply` or `thenCompose`.
 *
 * When we know about CompletableFutures that override `cancel` methods, we should convert them
 * to CancellableFuture before calling methods like `thenCompose`
 */
final class CancellableFuture[T](delegateFuture: CompletableFuture[T])
  extends CompletableFuture[T] {

  delegateFuture.whenComplete {
    case (value, null) => complete(value)
    case (_, ex) => completeExceptionally(ex)
  }

  override def cancel(mayInterruptIfRunning: Boolean): Boolean = {
    delegateFuture.cancel(mayInterruptIfRunning)
    super.cancel(mayInterruptIfRunning)
  }

  /**
   * * In Java 9, the new CompletableFuture's method was introduced: `newIncompleteFuture`.
   * * When using it, the cancel method of the original future will be called event when the
   * * `cancel` is done on a transformed one.
   */
  //override def newIncompleteFuture[U] = new CancellableFuture(delegateFuture)
}

File: bsp-builtin/bsp/src/org/jetbrains/bsp/protocol/session/CancellableFuture.java
package org.jetbrains.bsp.protocol.session;

import java.util.concurrent.CompletableFuture;

/**
 * JDK's CompletableFuture does not handle cancellation well.
 * When a future overrides a `cancel` method, it is lost
 * when doing transformations like `thenApply` or `thenCompose`.
 *
 * In Java 9, the new CompletableFuture's method was introduced: `newIncompleteFuture`.
 * When using it, the cancel method of the original future will be called event when the
 * `cancel` is done on a transformed one.
 *
 * When we know about CompletableFutures that override `cancel` methods, we should convert them
 * to CancellableFuture before calling methods like `thenCompose`
 *
 * The `newIncompleteFuture` is not annotated with @Override, so the code can be compiled with both JVM8 and JVM9
 * (and newer), but the actuall cancellation effect takes place only with 9 or newer.
 */
public class CancellableFuture<T> extends CompletableFuture<T> {

    final private CompletableFuture<?> original;

    public CancellableFuture(CompletableFuture<?> original)
    {
        this.original = original;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        original.cancel(mayInterruptIfRunning);
        return super.cancel(mayInterruptIfRunning);
    }

    @Override
    public <U> CompletableFuture<U> newIncompleteFuture() {
        return new CancellableFuture<>(original);
    }

    public static <U> CancellableFuture<U> from(CompletableFuture<U> original) {
        CancellableFuture<U> result = new CancellableFuture<>(original);
        original.whenComplete((value, error) -> {
            if(error != null) {
                result.completeExceptionally(error);
            } else {
                result.complete(value);
            }
        });
        return result;
    }
}

